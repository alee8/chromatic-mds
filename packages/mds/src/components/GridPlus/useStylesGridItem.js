import { createUseStyles } from 'react-jss';

import { BOX_SIZING_VALUES } from '~mds/styles/shared/css';
import { CSS_100 } from '~mds/styles/shared';
import { GRID_SIZE_VALUES } from '~mds/constants/sizes';

import { MAX_WIDTH_GRID_SYSTEM } from './constants';

/**
 * Amount of gutter spacing for a Wrench grid-item. 0 if size is not one of GRID_SIZE_OPTIONS.
 * @param {Object} grid Describes Wrench's grid
 * @param {String} size One of GRID_SIZE_OPTIONS
 * @returns size of gutter in Wrench grid item
 */
function gridGutterItem(grid, size) {
  return size ? grid.gutter[size] / 2 : 0;
}

/**
 *
 * @param {Float | Number | String} mWidth Requested maxWidth expressed as value between [0, 1.0] (float), width (number in terms of pixels) % (string)
 * @param {Number} breakpointWidth Given breakpoint's maxWidth
 * @returns
 */
function maxWidth(mWidth, breakpointWidth) {
  if (!mWidth) {
    // Ensure before any arithmetic, that NaN is produced because of undefined/null in the math
    // While the fix means that react-jss is now handling things correctly (ignoring NaN),
    // best for code not to even produce NaN. https://github.com/cssinjs/jss/issues/1445
    return undefined;
  }
  if (typeof mWidth === 'string') {
    // mWidth as string - as %
    return mWidth;
  }
  if (mWidth > 0 && mWidth <= 1.0) {
    // mWidth as float between [0, 1.0]
    return `${100 * mWidth}%`;
  }

  // mWidth is a number.  When it is 0, return
  if (typeof breakpointWidth === 'string') {
    // Legacy when breakpointWidth was based on Wrench 0.x.x had px in the breakpoint value
    return mWidth <= parseInt(breakpointWidth, 10) ? mWidth : undefined;
  }
  return mWidth <= breakpointWidth ? mWidth : undefined;
}

/**
 * Wrench's grid-item's width. 0 if size is not one of GRID_SIZE_OPTIONS
 * @param {Object} grid Describes Wrench's grid
 * @param {Number} numColumns A value between 0 and MAX_NUM_GRIDS
 * @returns width of Wrench grid item
 */
function width(grid, numColumns) {
  if (!numColumns) {
    // Ensure before any arithmetic, that NaN is produced because of undefined/null in the math
    // While the fix means that react-jss is now handling things correctly (ignoring NaN),
    // best for code not to even produce NaN. https://github.com/cssinjs/jss/issues/1445
    return undefined;
  }
  return `${100 * (numColumns / grid.columns)}%`;
}

const useStylesGridItem = createUseStyles((theme) => {
  const { breakpoints, mds: { grid } } = theme;

  return {
    mdsGridPlusItem: {
      boxSizing: BOX_SIZING_VALUES.borderBox,
      flexBasis: 0,
      flexGrow: 1,
      maxWidth: CSS_100,
      [`@media (min-width: ${breakpoints.xs}px)`]: {
        flex: ({ xs }) => xs && `0 0 ${width(grid, xs)}`,
        maxWidth: ({ xs, xsMaxWidth }) => maxWidth(xsMaxWidth, breakpoints.xs) ?? width(grid, xs),
        order: ({ xsOrder }) => xsOrder,
        paddingLeft: ({ noGutters }) => !noGutters && gridGutterItem(grid, GRID_SIZE_VALUES.xs),
        paddingRight: ({ noGutters }) => !noGutters && gridGutterItem(grid, GRID_SIZE_VALUES.xs),
      },

      [`@media (min-width: ${breakpoints.sm}px)`]: {
        flex: ({ sm }) => sm && `0 0 ${width(grid, sm)}`,
        maxWidth: ({ sm, smMaxWidth }) => maxWidth(smMaxWidth, breakpoints.sm) ?? width(grid, sm),
        order: ({ smOrder }) => smOrder,
        paddingLeft: ({ noGutters }) => !noGutters && gridGutterItem(grid, GRID_SIZE_VALUES.sm),
        paddingRight: ({ noGutters }) => !noGutters && gridGutterItem(grid, GRID_SIZE_VALUES.sm),
      },

      [`@media (min-width: ${breakpoints.md}px)`]: {
        flex: ({ md }) => md && `0 0 ${width(grid, md)}`,
        maxWidth: ({ md, mdMaxWidth }) => maxWidth(mdMaxWidth, breakpoints.md) ?? width(grid, md),
        order: ({ mdOrder }) => mdOrder,
        paddingLeft: ({ noGutters }) => !noGutters && gridGutterItem(grid, GRID_SIZE_VALUES.md),
        paddingRight: ({ noGutters }) => !noGutters && gridGutterItem(grid, GRID_SIZE_VALUES.md),
      },

      [`@media (min-width: ${breakpoints.lg}px)`]: {
        flex: ({ lg }) => lg && `0 0 ${width(grid, lg)}`,
        maxWidth: ({ lg, lgMaxWidth }) => maxWidth(lgMaxWidth, breakpoints.lg) ?? width(grid, lg),
        order: ({ lgOrder }) => lgOrder,
        paddingLeft: ({ noGutters }) => !noGutters && gridGutterItem(grid, GRID_SIZE_VALUES.lg),
        paddingRight: ({ noGutters }) => !noGutters && gridGutterItem(grid, GRID_SIZE_VALUES.lg),
      },

      [`@media (min-width: ${breakpoints.xl}px)`]: {
        flex: ({ xl }) => xl && `0 0 ${width(grid, xl)}`,
        maxWidth: ({ xl, xlMaxWidth }) => maxWidth(xlMaxWidth, MAX_WIDTH_GRID_SYSTEM) ?? width(grid, xl),
        order: ({ xlOrder }) => xlOrder,
        paddingLeft: ({ noGutters }) => !noGutters && gridGutterItem(grid, GRID_SIZE_VALUES.xl),
        paddingRight: ({ noGutters }) => !noGutters && gridGutterItem(grid, GRID_SIZE_VALUES.xl),
      },
    },
  };
});

export const TESTS = { maxWidth };
export default useStylesGridItem;
