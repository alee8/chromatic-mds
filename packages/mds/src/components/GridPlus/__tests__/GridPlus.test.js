/**
 * @jest-environment jsdom
 */

import React from 'react';
import { render } from '@testing-library/react';

import ComponentWrappers from '~tests/setupTests/ComponentWrappers';

import {
  Basic,
} from './Examples';

import { TESTS } from '../useStylesGridItem';

const TEST_OBJECT_PROP_OPTIONS = [
  { width: 0.3, size: '', expected: '30%' },
  { width: 476, size: 768, expected: 476 },
  { width: 998, size: 768, expected: undefined },
  { width: 0.90, size: 768, expected: '90%' },
  { width: 0, size: 768, expected: undefined },
  { width: 1, size: 768, expected: '100%' },
  { width: 1.0, size: 768, expected: '100%' },
  { width: 'calc(100% - 2 * 24px)', size: 768, expected: 'calc(100% - 2 * 24px)' },
];

describe('Test maxWidth function in GridPlusItem', () => {
  test('test maxwidth with different values', () => {
    TEST_OBJECT_PROP_OPTIONS.forEach(
      (test) => {
        expect(TESTS.maxWidth(test.width, test.size)).toBe(test.expected);
      }
    );
  });
});

describe('Render GridPlus Snapshot', () => {
  test('Basic GridPlus', () => {
    const { container, unmount } = render(
      <ComponentWrappers>
        <Basic />
      </ComponentWrappers>
    );

    expect(container).toMatchSnapshot();
    unmount(container);
  });
});
