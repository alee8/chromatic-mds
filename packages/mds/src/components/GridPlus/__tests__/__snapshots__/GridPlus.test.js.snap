// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render GridPlus Snapshot Basic GridPlus 1`] = `
<div>
  <div
    class="mntv-bodyBackground-0-2-1 mntv-wrenchBox-0-2-6 mntv-wrenchBox-d0-0-2-7 mntv-wrenchSpacing-0-2-4 mntv-wrenchSpacing-d0-0-2-5"
  >
    <div
      class="mntv-wrenchBox-0-2-6 mntv-wrenchBox-d1-0-2-9 mntv-wrenchSpacing-0-2-4 mntv-wrenchSpacing-d1-0-2-8"
    >
      <div
        class="noneMarginBottom mdMarginTop nonePaddingBottom nonePaddingTop mntv-heading-0-2-10 mntv-heading-d0-0-2-13  mntv-headingSize-0-2-11 mntv-headingSize-d1-0-2-14  "
        id=""
      >
        <h3>
          GridPlus props
        </h3>
      </div>
      <p
        class="mntv-background-0-2-15 mntv-background-d1-0-2-19  mntv-dropCap-0-2-16 mntv-dropCap-d4-0-2-20  mntv-paragraph-0-2-17 mntv-paragraph-d5-0-2-21  mntv-paragraphTypography-0-2-18 mntv-paragraphTypography-d6-0-2-22  "
      >
        Yellow highlights fluid area when fluid=false
      </p>
      <p
        class="mntv-background-0-2-15 mntv-background-d9-0-2-23  mntv-dropCap-0-2-16 mntv-dropCap-d12-0-2-24  mntv-paragraph-0-2-17 mntv-paragraph-d13-0-2-25  mntv-paragraphTypography-0-2-18 mntv-paragraphTypography-d14-0-2-26  "
      >
        Orange highlights bleed area when noBleeds=false
      </p>
      <p
        class="mntv-background-0-2-15 mntv-background-d17-0-2-27  mntv-dropCap-0-2-16 mntv-dropCap-d20-0-2-28  mntv-paragraph-0-2-17 mntv-paragraph-d21-0-2-29  mntv-paragraphTypography-0-2-18 mntv-paragraphTypography-d22-0-2-30  "
      >
        Green highlights gutter areas when noGutters=false
      </p>
    </div>
    <div
      class="mntv-fullHighlight-0-2-3 mntv-wrenchBox-0-2-6 mntv-wrenchBox-d2-0-2-32 mntv-wrenchSpacing-0-2-4 mntv-wrenchSpacing-d2-0-2-31"
    >
      <div
        class="mntv-mdsGridPlusContainer-0-2-35 mntv-mdsGridPlusContainer-d5-0-2-37"
      >
        <div
          class="mntv-mdsGridPlus-0-2-36 mntv-mdsGridPlus-d11-0-2-38 mntv-addFlex-0-2-33 mntv-addFlex-d0-0-2-34"
        >
          <div
            class="mntv-mdsGridPlusItem-0-2-41 mntv-mdsGridPlusItem-d0-0-2-42 mntv-addFlex-0-2-39 mntv-addFlex-d0-0-2-40"
          >
            <div
              style="background-color: rgb(208, 208, 208); border: 2px dashed #9d9d9d; padding: 16px; font-family: Roboto, Helvetica, Arial;"
            >
              Item 0
            </div>
          </div>
          <div
            class="mntv-mdsGridPlusItem-0-2-41 mntv-mdsGridPlusItem-d6-0-2-44 mntv-addFlex-0-2-39 mntv-addFlex-d1-0-2-43"
          >
            <div
              style="background-color: rgb(208, 208, 208); border: 2px dashed #9d9d9d; padding: 16px; font-family: Roboto, Helvetica, Arial;"
            >
              Item 1
            </div>
          </div>
          <div
            class="mntv-mdsGridPlusItem-0-2-41 mntv-mdsGridPlusItem-d12-0-2-46 mntv-addFlex-0-2-39 mntv-addFlex-d2-0-2-45"
          >
            <div
              style="background-color: rgb(208, 208, 208); border: 2px dashed #9d9d9d; padding: 16px; font-family: Roboto, Helvetica, Arial;"
            >
              Item 2
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
